File: 2019/cpp/day_01b.cpp
#include <fstream>
#include <iostream>
#include <iterator>
#include <string>
#include <vector>

int getRecursiveFuel(const int mass) {
  constexpr int divisor = 3;
  constexpr int subtract = 2;
  constexpr int minimum_mass = divisor * subtract;
  if (mass < minimum_mass) {
    return 0;
  }
  const int fuel = mass / divisor - subtract;
  return fuel + getRecursiveFuel(fuel);
}

int main(int argc, char* argv[]) {
  // Get input
  std::string input = "../input/day_01_input";
  if (argc > 1) {
    input = argv[1];
  }
  std::ifstream file(input);
  std::istream_iterator<int> start(file), end;
  const std::vector<int> masses(start, end);

  // Solve
  long long total_fuel = 0;
  for (const auto& mass : masses) {
    total_fuel += getRecursiveFuel(mass);
  }
  std::cout << total_fuel << '\n';
  return total_fuel;
}

File: 2019/python/day_01b.py
#!/usr/bin/python3

import sys


def getRecursiveFuel(mass):
    divisor = 3
    subtract = 2
    minimum_mass = divisor * subtract
    if mass < minimum_mass:
        return 0
    fuel = int(mass / divisor) - subtract
    return fuel + getRecursiveFuel(fuel)


def main():
    # Get input
    file_name = "../input/day_01_input"
    if len(sys.argv) > 1:
        file_name = sys.argv[1]
    file = open(file_name)
    input = [int(ele) for ele in file]

    # Solve
    total_fuel = 0
    for mass in input:
        total_fuel = total_fuel + getRecursiveFuel(mass)
    print(total_fuel)
    return total_fuel


if __name__ == "__main__":
    main()
